source/common/http/http2/codec_impl.h

```diff
@@ -187,11 +187,16 @@ class ConnectionImpl : public virtual Connection, protected Logger::Loggable<Log
     // I don't fully understand.
     static const uint64_t MAX_HEADER_SIZE = 63 * 1024;
 
+    // Does any necessary WebSocket/Upgrade conversion, then passes the headers
+    // to the decoder_.
+    void decodeHeaders();
+
     bool buffers_overrun() const { return read_disable_count_ > 0; }
 
     ConnectionImpl& parent_;
     HeaderMapImplPtr headers_;
     StreamDecoder* decoder_{};
+    std::string upgrade_type_;
```

Comment:
What do you think of qualifying this with websocket alone? Or may be an enum for other protos in future? The reason is that Websocket is the only Http-ish proto in this picture. Others are going to be just opaque TCP.